extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/OidcClientConfigBuilder.java

```diff
@@ -0,0 +1,395 @@
+    }
+
+    public static final class GrantBuilder {
+
+        private record GrantImpl(Type type, String accessTokenProperty, String refreshTokenProperty, String expiresInProperty,
+                String refreshExpiresInProperty) implements Grant {
+
+        }
+
+        private final OidcClientConfigBuilder builder;
+        private Grant.Type type;
+        private String accessTokenProperty;
+        private String refreshTokenProperty;
+        private String expiresInProperty;
+        private String refreshExpiresInProperty;
+
+        public GrantBuilder() {
+            this.builder = null;
+        }
+
+        public GrantBuilder(OidcClientConfigBuilder builder) {
+            this.builder = builder;
+            this.type = builder.grant.type();
+            this.accessTokenProperty = builder.grant.accessTokenProperty();
+            this.refreshTokenProperty = builder.grant.refreshTokenProperty();
+            this.expiresInProperty = builder.grant.expiresInProperty();
+            this.refreshExpiresInProperty = builder.grant.refreshExpiresInProperty();
+        }
+
+        /**
+         * @param type {@link Grant#type()}
+         * @return this builder
+         */
+        public GrantBuilder type(Grant.Type type) {
+            this.type = Objects.requireNonNull(type);
+            return this;
+        }
+
+        /**
+         * @param refreshTokenProperty {@link Grant#refreshTokenProperty()}
+         * @return this builder
+         */
+        public GrantBuilder refreshTokenProperty(String refreshTokenProperty) {
+            this.refreshTokenProperty = Objects.requireNonNull(refreshTokenProperty);
+            return this;
+        }
+
+        /**
+         * @param expiresInProperty {@link Grant#expiresInProperty()}
+         * @return this builder
+         */
+        public GrantBuilder expiresInProperty(String expiresInProperty) {
+            this.expiresInProperty = Objects.requireNonNull(expiresInProperty);
+            return this;
+        }
+
+        /**
+         * @param refreshExpiresInProperty {@link Grant#refreshExpiresInProperty()}
+         * @return this builder
+         */
+        public GrantBuilder refreshExpiresInProperty(String refreshExpiresInProperty) {
+            this.refreshExpiresInProperty = Objects.requireNonNull(refreshExpiresInProperty);
+            return this;
+        }
+
+        /**
+         * @param accessTokenProperty {@link Grant#accessTokenProperty()}
+         * @return this builder
+         */
+        public GrantBuilder accessTokenProperty(String accessTokenProperty) {
+            this.accessTokenProperty = Objects.requireNonNull(accessTokenProperty);
+            return this;
+        }
+
+        public OidcClientConfigBuilder end() {
+            Objects.requireNonNull(builder);
+            return builder.grant(build());
+        }
+
+        public Grant build() {
+            Objects.requireNonNull(type, "OIDC Client grant type must be set");
```

Comment:
But that makes some optional properties non-optional, or am I'm missing something ?

Reply:
I think I can explain that:

- none of these properties is optional
- if user do what I suggest (typing it out of my head, maybe it is not one to one to code):
  - that is `var grantBuilder = OidcClientConfigBuilder.builder().grant()`
  - or `var builder = OidcClientConfigBuilder.builder();var grantBuilder = new GrantBuilder(builder);` then the exceptions you are commenting on will never be thrown, no way
  - but if users do just `var grantBuilder = new GrantBuilder();` then what are our options?
    - [ref. number: 123] we can build config and get defaults from the SmallRye Config, we can do that, but it's less efficient because you are building config again
    - we can have hardcoded defaults (NO WAY, it's bad for keeping things consistent)\

So, yes, I can avoid that by [123], but I thought that this will motivate users to the other options (use existing builder).

If you want, I'll fix it. Please let me know.

Reply:
@michalvavrik Sorry, I forgot they all have default values, so when `var grantBuilder = new GrantBuilder();` then I believe we should set properties to the same default values that are set in `ConfigMapping`, that would require us keeping an eye to make sure they don't go out of sync, but IMHO it is better than asking users to set an `expiredInProperty` when they work with standard OAuth with a well known expires in property

Reply:
ok, I'll have a look

Reply:
@sberyozkin solution was quite easy, I dropped this, now this `GrantBuilder` is always populated with defaults. Additionally I have added a test to test that's the case